File: Offline-Playback/app/src/main/java/com/theoplayer/sample/playback/offline/SourceDescriptionRepository.java
package com.theoplayer.sample.playback.offline;

import android.content.Context;

import com.theoplayer.android.api.source.SourceDescription;
import com.theoplayer.android.api.source.drm.LicenseType;

import static com.theoplayer.android.api.source.SourceDescription.Builder.sourceDescription;
import static com.theoplayer.android.api.source.TypedSource.Builder.typedSource;
import static com.theoplayer.android.api.source.drm.DRMConfiguration.Builder.widevineDrm;
import static com.theoplayer.android.api.source.drm.KeySystemConfiguration.Builder.keySystemConfiguration;

public final class SourceDescriptionRepository {

    public static SourceDescription getBySourceUrl(Context context, String sourceUrl) {
        if (sourceUrl.equals(context.getString(R.string.bigBuckBunnySourceUrl))) {
            return getBigBuckBunnySourceDescription(context);
        }
        if (sourceUrl.equals(context.getString(R.string.sintelSourceUrl))) {
            return getSintelSourceDescription(context);
        }
        if (sourceUrl.equals(context.getString(R.string.tearsOfStealSourceUrl))) {
            return getTearsOfSteelSourceDescription(context);
        }
        if (sourceUrl.equals(context.getString(R.string.elephantsDreamSourceUrl))) {
            return getElephantsDreamSourceDescription(context);
        }
        return null;
    }

    private static SourceDescription getBigBuckBunnySourceDescription(Context context) {
        return sourceDescription(
                typedSource(context.getString(R.string.bigBuckBunnySourceUrl))
                        .drm(
                                widevineDrm(
                                        // Note that license has to have PERSISTENT type configured
                                        // to be cached and to allow offline playback.
                                        keySystemConfiguration(context.getString(R.string.bigBuckBunnyLicenseUrl))
                                                .licenseType(LicenseType.PERSISTENT)
                                                .build()
                                ).build()
                        )
                        .setExperimentalRenderingEnabled(true)
                        .build()
        ).build();
    }

    private static SourceDescription getSintelSourceDescription(Context context) {
        return sourceDescription(
                typedSource(context.getString(R.string.sintelSourceUrl)).build()
        ).build();
    }

    private static SourceDescription getTearsOfSteelSourceDescription(Context context) {
        return sourceDescription(
                typedSource(context.getString(R.string.tearsOfStealSourceUrl)).build()
        ).build();
    }

    private static SourceDescription getElephantsDreamSourceDescription(Context context) {
        return sourceDescription(
                typedSource(context.getString(R.string.elephantsDreamSourceUrl)).build()
        ).build();
    }

}

File: offline-playback/src/main/java/com/theoplayer/sample/playback/offline/SourceDescriptionRepository.kt
package com.theoplayer.sample.playback.offline

import android.content.Context
import com.theoplayer.android.api.source.SourceDescription
import com.theoplayer.android.api.source.TypedSource
import com.theoplayer.android.api.source.drm.DRMConfiguration
import com.theoplayer.android.api.source.drm.KeySystemConfiguration
import com.theoplayer.android.api.source.drm.LicenseType

object SourceDescriptionRepository {
    fun getBySourceUrl(context: Context, sourceUrl: String?): SourceDescription? =
        when (sourceUrl) {
            context.getString(R.string.bigBuckBunnySourceUrl) -> getBigBuckBunnySourceDescription(
                context
            )

            context.getString(R.string.sintelSourceUrl) -> getSintelSourceDescription(context)
            context.getString(R.string.tearsOfStealSourceUrl) -> getTearsOfSteelSourceDescription(
                context
            )

            context.getString(R.string.bipBopSourceUrl) -> getBipBopSourceDescription(
                context
            )

            else -> null
        }

    private fun getBigBuckBunnySourceDescription(context: Context): SourceDescription {
        return SourceDescription.Builder(
            TypedSource.Builder(context.getString(R.string.bigBuckBunnySourceUrl))
                .build()
        ).build()
    }

    private fun getSintelSourceDescription(context: Context): SourceDescription {
        return SourceDescription.Builder(
            TypedSource.Builder(
                context.getString(R.string.sintelSourceUrl)
            ).build()
        ).build()
    }

    private fun getTearsOfSteelSourceDescription(context: Context): SourceDescription {
        return SourceDescription.Builder(
            TypedSource.Builder(
                context.getString(R.string.tearsOfStealSourceUrl)
            ).drm(
                DRMConfiguration.Builder()
                    .widevine( // Note that license has to have PERSISTENT type configured
                        // to be cached and to allow offline playback.
                        KeySystemConfiguration.Builder(context.getString(R.string.tearsOfStealLicenseUrl))
                            .licenseType(LicenseType.PERSISTENT)
                            .build()
                    ).build()
            ).build()
        ).build()
    }

    private fun getBipBopSourceDescription(context: Context): SourceDescription {
        return SourceDescription.Builder(
            TypedSource.Builder(
                context.getString(R.string.bipBopSourceUrl)
            ).build()
        ).build()
    }
}